Pass over alert section (#631)
browserengineering/book: book/accessibility.md

:::

> Live Regions

Reviewer: ```suggestion
Accessibility alerts
```

============

Scripts do not interact directly with the accessibility tree, much
like they do not interact directly with the display list. However,
sometimes scripts need to inform the screen reader about *why* they're
making certain changes to the page to give screen-reader users a
better experience. The most common example is an alert[^toast] telling

> you that some action you just did failed. A screen-reader user needs

Reviewer: ```suggestion
you that some action you just did failed. A screen reader user needs
```

the alert read to them immediately, no matter where in the document
it's inserted.

[^toast]: Also called a "toast", because it pops up.

> The `alert` role addresses this need. A screen-reader will

Reviewer: ```suggestion
The `alert` role addresses this need. A screen reader will
```

immediately[^alert-css] read an element with that role, no matter
where in the document the user currently is. Note that there aren't
any HTML elements whose default role is `alert`, so this requires

> setting the `role` attribute. There are also other "live" roles like

Reviewer: Put this last sentence in a footnote perhaps?

:::

Live Regions
============

Scripts do not interact directly with the accessibility tree,[^aom]
much like they do not interact directly with the display list.
However, sometimes scripts need to inform the screen reader about
*why* they're making certain changes to the page to give screen-reader
users a better experience. The most common example is an alert[^toast]
telling you that some action you just did failed. Typically, these are
inserted at some fixed place in the document, and thus in the

> accessibility tree; there's little chance a screen-reader user would

Reviewer: The second half of this sentence is confusing. I think you're trying to say that you can't just rely on the regular document-reading order for alerts, but it's not clear enough.

get to it in time to see it.

> [^aom]: As of this writing, there's a draft [Accessibility Objet

Reviewer: This was abandoned, you should just remove it.

    Model](https://wicg.github.io/aom/explainer.html) specification,
    which would allow scripts to directly create accessibility nodes.
    
[^toast]: Also called a "toast", because it pops up.

To fix this, if an element's role is set to `alert`, the screen-reader
will immediately[^alert-css] read that element without the user
needing move around in the accessiblity tree to find it. Note that
there aren't any HTML elements whose default role is `alert`, so this
requires setting the `role` attribute. There are also other "live"
roles like `status` for less urgent information or `alertdialog` if
the keyboard focus should move to the alerted element.

[^alert-css]: The alert is only triggered if the element is added to
    the document, has the `alert` role (or the equivalent `aria-live`
    value, `assertive`), and is visible in the layout tree (meaning it
    doesn't have `display: none`), or if its contents change. In this
    chapter, I won't handle all of these cases and just focus on new
    elements with an `alert` role, not changes to contents or CSS.
    
Before we jump to implementation, we first need to make it possible

> for scripts to change the `role` attribute. Let's add support for the

Reviewer: ```suggestion
for scripts to change the `role` attribute. To do that we'll need to add support for the
```

    def setAttribute(self, handle, attr, value):
        elt = self.handle_to_node[handle]
        elt.attributes[attr] = value
```

Now we can implement the `alert` role. To do so, we'll search the

> accessiblity tree for elements with the `alert` role:

Reviewer: ```suggestion
accessiblity tree for elements with that role:
```

``` {.python}
class Tab:
    def __init__(self, browser):
        # ...
        self.active_alerts = []

    def render(self):
        if self.needs_accessibility:
            # ...
            self.active_alerts = [
                node for node in tree_to_list(self.accessibility_tree, [])
                if node.role == "alert"
            ]
```

Now, we can't just read out every `alert` at every frame; we need to
keep track of what elements have already been read, so we don't read
them twice:

``` {.python}
class Tab:
    def __init__(self, browser):
        # ...
        self.spoken_alerts = []

    def speak_update(self):
        for alert in self.active_alerts:
            if alert not in self.spoken_alerts:

>                 self.speak_node(alert, "New alert")

Reviewer: perhaps add an exercise to allow setting the text content of an element in order to update the alert. Otherwise the page will have to remove the alert, wait a rAF loop, then re-add it with new innerHTML.

[hover-pseudo]: https://developer.mozilla.org/en-US/docs/Web/CSS/:hover

* *Alert updates*: Right now, the screen-reader immediately reads an
alert when an element gains the `alert` role. But if that element's
contents change, the alert isn't re-read. In real browsers, changes to
an alert cause the alert to be re-read. Implement that in your
browser: if the text description of any node below an `alert` node

> changes in the accessibility tree, re-read the `alert`. For an

Reviewer: I don't think our runtime currently supports changing out the text without the element...also add that API?

Author: You can change the `innerHTML` of an existing element to regular text. It won't change the element itself, just delete existing children and add a new one.

Reviewer: Right, forgot about that.
